### DISCUSSION TITLE
Uncaught Snapshot missing on Livewire component with id: ...

### DISCUSSION THREAD

**Discussion by pboese** [livewire/livewire#7602] [1. Help]
Heya, I'm pretty new to Livewire (using v3.3.3) and am having an issue with a simple component that uses two embedded (child) components. Here's my component:

```html
<div>
    <x-card :title="__('Neuer Kontakt')" shadow separator progress-indicator>
        <x-form wire:submit="save">
            <div class="grid grid-cols-2 gap-4">
                <div><x-input wire:model="first_name" label="{{ __('Vorname') }} *" /></div>
                <div><x-input wire:model="last_name" label="{{ __('Nachname') }} *" /></div>

                <div><x-input wire:model="phone" label="{{ __('Telefon') }}" /></div>
                <div><x-input wire:model="email" label="{{ __('E-Mail') }}" type="email" /></div>

                <div class="col-span-2 text-right"><x-button type="submit" spinner="save" class="btn-primary">@lang('Speichern')</x-button></div>
            </div>
        </x-form>
    </x-card>

    <x-card>
        <x-tabs selected="seller_buyer-tab">
            <x-tab name="seller_buyer-tab" :label="__('Käufer / Verkäufer')">
                <livewire:contact-table category="seller_buyer" />
            </x-tab>

            <x-tab name="service_provider-tab" :label="__('Dienstleister')">
                <livewire:contact-table category="service_provider" />
            </x-tab>
        </x-tabs>
    </x-card>
</div>
```

Whenever I submit the form in the component, the two child components (`<livewire:contact-table/>`) refresh and this results in something like the following errors:

```
Uncaught Snapshot missing on Livewire component with id: nK6MlBO6TcSOXVm6Fqju
Uncaught Snapshot missing on Livewire component with id: yi8xN87a2KqwEk2y5TrP
```

I tried adding `wire:key` on various elements but I'm doing something wrong it seems. Any hints are very welcome :-) Thanks in advance!

**Comment by lrljoe** [comment]
The common gotchas are when you don't:
- Add unique keys to your components
- Have something odd in the route of your nested component

Are your tables and cards uniquely identified?  You've got two of the same component there, both from a Component and Livewire Component element so make sure they're all keyed uniquely.

**Comment by devajmeireles** [comment]
> Heya, I'm pretty new to Livewire (using v3.3.3) and am having an issue with a simple component that uses two embedded (child) components. Here's my component:
> 
> ```
> <div>
>     <x-card :title="__('Neuer Kontakt')" shadow separator progress-indicator>
>         <x-form wire:submit="save">
>             <div class="grid grid-cols-2 gap-4">
>                 <div><x-input wire:model="first_name" label="{{ __('Vorname') }} *" /></div>
>                 <div><x-input wire:model="last_name" label="{{ __('Nachname') }} *" /></div>
> 
>                 <div><x-input wire:model="phone" label="{{ __('Telefon') }}" /></div>
>                 <div><x-input wire:model="email" label="{{ __('E-Mail') }}" type="email" /></div>
> 
>                 <div class="col-span-2 text-right"><x-button type="submit" spinner="save" class="btn-primary">@lang('Speichern')</x-button></div>
>             </div>
>         </x-form>
>     </x-card>
> 
>     <x-card>
>         <x-tabs selected="seller_buyer-tab">
>             <x-tab name="seller_buyer-tab" :label="__('Käufer / Verkäufer')">
>                 <livewire:contact-table category="seller_buyer" />
>             </x-tab>
> 
>             <x-tab name="service_provider-tab" :label="__('Dienstleister')">
>                 <livewire:contact-table category="service_provider" />
>             </x-tab>
>         </x-tabs>
>     </x-card>
> </div>
> ```
> 
> Whenever I submit the form in the component, the two child components (`<livewire:contact-table/>`) refresh and this results in something like the following errors:
> 
> ```
> Uncaught Snapshot missing on Livewire component with id: nK6MlBO6TcSOXVm6Fqju
> Uncaught Snapshot missing on Livewire component with id: yi8xN87a2KqwEk2y5TrP
> ```
> 
> I tried adding `wire:key` on various elements but I'm doing something wrong it seems. Any hints are very welcome :-) Thanks in advance!

@pboese , I've been through something similar in the past. Over time I discovered that apparently, Livewire has a certain difference between `wire:key` and `:key` (I'm not sure about it!). In my case, the problem was only solved by using `:key` instead of `wire:key`

**Comment by pboese** [comment]
Thanks a lot @irqd and @devajmeireles! Seems the issue is https://github.com/robsontenorio/mary/issues/243

**Comment by jefpes** [comment]
deal with it

before (Uncaught Snapshot missing on Livewire component with id: YTCHpBfqn19t6rdMqki7):

         @foreach ($this->roles as $r)
          <div :key="'role_row_' . $r->id">
            <x-table.tr >
              <x-table.td> {{ $r->name }} </x-table.td>
              <x-table.td> {{ $r->hierarchy }} </x-table.td>
              <x-table.td>
                @if (auth()->user()->roles()->pluck('hierarchy')->max() <= $r->hierarchy)
                  <div class="flex flex-row gap-2 justify-center">
                    <livewire:role.edit :id="$r->id" :key="'role_edit_' . $r->id"/>
    
                    <a id="ability_role_button_{{ $r->id }}" href="ability-roles/{{ $r->id }}" wire:navigate>
                      <x-icons.shield  class="cursor-pointer flex text-green-600 w-8 h-8"/>
                    </a>
    
                    <livewire:role.delete :id="$r->id" :key="'role_delete_' . $r->id"/>
                  </div>
                @endif
              </x-table.td>
            </x-table.tr>
          </div>
        @endforeach
        
after:

        @foreach ($this->roles as $r)
          <livewire:role.row :id="$r->id" :key="$r->id" />
        @endforeach
        
role.row:

    <x-table.tr>
      <x-table.td> {{ $form->name }} </x-table.td>
      <x-table.td> {{ $form->hierarchy }} </x-table.td>
      <x-table.td>
        @if (auth()->user()->roles()->pluck('hierarchy')->max() <= $form->hierarchy)
         <div class="flex flex-row gap-2 justify-center">
             <livewire:role.edit :id="$form->id" wire:key="edit_role_button_{{ $form->id }}"/>

            <a id="ability_role_button_{{ $form->id }}" href="ability-roles/{{ $form->id }}" wire:navigate> <x-icons.shield  class="cursor-pointer flex text-green-600 w-8 h-8"/> </a>

            <livewire:role.delete :id="$form->id" wire:key="delete_role_button_{{ $form->id }}"/>
         </div>
        @endif
      </x-table.td>
    </x-table.tr>

**Reply by lrljoe** [reply]
So long as thats a UUID, you can use it as a key, I'd reconmend not using a key that is just an integer, as you ultimately may end up with two items with the same key
e.g
A list of users with IDs of 1-10
A list of addresses with IDs of 1-10

Instead prefix your keys with something unique if you're not using uuids as your primary key

**Comment by cbaconnier** [comment]
I do have some times to times this error from the error reporting tools.
Any idea how I could catch that error so could add some context and refresh the page ?

In my case it's a monitoring tool, so catching that error to refresh the page would also solve the issue.


--

Edit:
This would work for me
```js
window.addEventListener('error', function(error) {
    if (error instanceof ErrorEvent) {
        if (error.message.includes('Snapshot missing on Livewire component with id:')) {
            // report the error
            // ...

            // refresh the page. (Using 5 second  delay to avoid an accidental self DDoS)
            setTimeout(function() {
                location.reload();
            }, 5000)
        }
    }
});
```

**Comment by daviddierose** [comment]
This happened to me once as well. In my case, the issue was caused by a UTF-8 BOM at the beginning of the file. After removing the BOM and saving the file as UTF-8 without BOM, the Livewire errors disappeared.
